include/deal.II/matrix_free/evaluation_kernels.h

```diff
@@ -2165,10 +2204,126 @@ namespace internal
         });
     }
 
-    template <std::size_t n_face_orientations>
+    template <std::size_t n_face_orientations,
+              typename VectorType,
+              typename std::enable_if<
+                !internal::has_begin<VectorType>::value ||
+                  !std::is_same<decltype(std::declval<VectorType>().begin()),
+                                Number2 *>::value,
+                VectorType>::type * = nullptr>
+    static bool
+    integrate_scatter(
+      VectorType &,
+      const MatrixFreeFunctions::ShapeInfo<VectorizedArrayType> &data,
+      const MatrixFreeFunctions::DoFInfo &,
+      VectorizedArrayType *values_array,
+      VectorizedArrayType *values_quad,
+      VectorizedArrayType *gradients_quad,
+      VectorizedArrayType *scratch_data,
+      const bool           integrate_values,
+      const bool           integrate_gradients,
+      const unsigned int,
+      const unsigned int,
+      const std::array<unsigned int, n_face_orientations>,
+      const std::array<unsigned int, n_face_orientations> face_nos,
+      const unsigned int                                  subface_index,
+      const MatrixFreeFunctions::DoFInfo::DoFAccessIndex,
+      const std::array<unsigned int, n_face_orientations> face_orientations,
+      const Table<2, unsigned int> &                      orientation_map)
+    {
```

Comment:
Can you explain why you need to add the code inside this function? At first sight I do not see the difference to `FEFaceEvaluationSelector::integrate` (well, except some arguments). At least, this would need some comments on why things get through another path than in the `evaluate` routine where we can bypass the operation completely.

Reply:
I guess I could call `integrate` in this function. The general question is where we want check if a vector class has the right `begin()`-method. I have chosen to specialize for `FEFaceEvaluationSelector::integrate_scatter()` because I found it more easy in this case; for `FEEvaluation::integrate_scatter` it is done differently...

Reply:
I have updated this PR. How do you like it this way?

Reply:
> Excellent!

This PR has lead to some annoying incompatibilities in one of our application codes where we precompile (in one way or the other) these intenal kernels. To add additional template arguments here seems not to be a good idea. We should probably move the check to `FEFaceEvaluation`.

FYI @mschreter

Reply:
I agree, but I would append that we need frequent changes in code using these internal interfaces during these days anyway.